realm/kotlin-extensions/src/androidTest/kotlin/io/realm/CoroutineTests.kt

```diff
@@ -0,0 +1,510 @@
+        testDispatcher = TestCoroutineDispatcher()
+        testScope = TestCoroutineScope(testDispatcher)
+        configuration = configFactory.createConfiguration()
+        realm = Realm.getInstance(configuration)
+    }
+
+    @After
+    fun tearDown() {
+        realm.executeTransaction { it.deleteAll() }
+        realm.close()
+    }
+
+    @Test
+    fun toFlow_emittedOnCollect() {
+        val countDownLatch = CountDownLatch(1)
+
+        // TODO check this out for better testing: https://proandroiddev.com/from-rxjava-to-kotlin-flow-testing-42f1641d8433
+        val context = Dispatchers.Main
+        val scope = CoroutineScope(context)
+
+        scope.launch {
+            val realmInstance = Realm.getInstance(configuration)
+
+            realmInstance.where<SimpleClass>()
+                    .findAllAsync()
+                    .toFlow()
+                    .flowOn(context)
+                    .onEach { flowResults ->
+                        assertTrue(flowResults.isFrozen)
+                        assertEquals(0, flowResults.size)
+                        scope.cancel("Cancelling scope...")
+                    }.onCompletion {
+                        realmInstance.close()
+                        countDownLatch.countDown()
+                    }.launchIn(scope)
+        }
+
+        TestHelper.awaitOrFail(countDownLatch)
+    }
+
+    @Test
+    fun toFlow_resultsEmittedAfterCollect() {
+        realm.executeTransactionAsync { transactionRealm ->
+            transactionRealm.createObject<SimpleClass>().name = "Foo"
+            transactionRealm.createObject<SimpleClass>().name = "Bar"
+        }
+
+        val countDownLatch = CountDownLatch(1)
+
+        val context = Dispatchers.Main
+        val scope = CoroutineScope(context)
+
+        scope.launch {
+            val realmInstance = Realm.getInstance(configuration)
+            realmInstance.where<SimpleClass>()
+                    .findAllAsync()
+                    .toFlow()
+                    .flowOn(context)
+                    .onEach { flowResults ->
+                        assertTrue(flowResults.isFrozen)
+                        if (flowResults.size == 2) {
+                            scope.cancel("Cancelling scope...")
+                        }
+                    }.onCompletion {
+                        realmInstance.close()
+                        countDownLatch.countDown()
+                    }.launchIn(scope)
+        }
+
+        TestHelper.awaitOrFail(countDownLatch)
+    }
+
+    @Test
+    fun toFlow_resultsCancelBeforeCollect() {
+        val countDownLatch = CountDownLatch(1)
+
+        val context = Dispatchers.Main
+        val scope = CoroutineScope(context)
+
+        scope.launch {
+            val realmInstance = Realm.getInstance(configuration)
+            realmInstance.where<SimpleClass>()
+                    .findAllAsync()
+                    .toFlow()
+                    .flowOn(context)
+                    .onCompletion {
```

Comment:
`toFlow` emits whatever results we have immediately, and normally it is an empty result the first time, which means `onEach` will be called most likely at least once with empty `RealmResults`. However, since this test is tightly synchronised - look at the `delay(100)` just a few lines below - maybe it doesn't even make sense to test it...?